x/superfluid/keeper/epoch_test.go

```diff
@@ -106,3 +111,255 @@ func (s *KeeperTestSuite) TestUpdateOsmoEquivalentMultipliers() {
 		})
 	}
 }
+
+type gaugeChecker struct {
+	intermediaryAccIndex     uint64
+	valIndex                 int64
+	lockIndexes              []int64
+	lpIndex                  int64
+	rewarded                 bool
+	expectedDistributedCoins sdk.Coins
+}
+type distributionTestCase struct {
+	name             string
+	validatorStats   []stakingtypes.BondStatus
+	superDelegations []superfluidDelegation
+	rewardedVals     []int64
+	gaugeChecks      []gaugeChecker
+}
+
+var (
+	// distributed coin when there is one account receiving from one gauge
+	defaultSingleLockDistributedCoins = sdk.NewCoins(sdk.NewInt64Coin("stake", 19999))
+	// distributed coins when there is two account receiving from one gauge
+	defaultTwoLockDistributedCoins = sdk.NewCoins(sdk.NewInt64Coin("stake", 9999))
+	// distributed coins when there is one account receiving from two gauge
+	defaultTwoGaugeDistributedCoins = sdk.NewCoins(sdk.NewInt64Coin("stake", 19998))
+	distributionTestCases           = []distributionTestCase{
+		{
+			"happy path with single validator and delegator",
+			[]stakingtypes.BondStatus{stakingtypes.Bonded},
+			[]superfluidDelegation{{0, 0, 0, 1000000}},
+			[]int64{0},
+			[]gaugeChecker{{0, 0, []int64{0}, 0, true, defaultSingleLockDistributedCoins}},
+		},
+		{
+			"two LP tokens delegation to a single validator",
+			[]stakingtypes.BondStatus{stakingtypes.Bonded},
+			[]superfluidDelegation{{0, 0, 0, 1000000}, {0, 0, 1, 1000000}},
+			[]int64{0},
+			[]gaugeChecker{{0, 0, []int64{0}, 0, true, defaultTwoLockDistributedCoins}, {1, 0, []int64{0}, 1, true, defaultTwoLockDistributedCoins}},
+		},
+		{
+			"one LP token with two locks to a single validator",
+			[]stakingtypes.BondStatus{stakingtypes.Bonded},
+			[]superfluidDelegation{{0, 0, 0, 1000000}, {1, 0, 0, 1000000}},
+			[]int64{0},
+			[]gaugeChecker{{0, 0, []int64{0, 1}, 0, true, defaultTwoGaugeDistributedCoins}},
+		},
+		// In this case, allocate reward to validators with different stat.
+		// There is no difference between Bonded, Unbonding, Unbonded
+		{
+			"add unbonded, unbonding validator case",
+			[]stakingtypes.BondStatus{stakingtypes.Bonded, stakingtypes.Unbonded, stakingtypes.Unbonding},
+			[]superfluidDelegation{{0, 0, 0, 1000000}, {1, 1, 0, 1000000}, {2, 2, 0, 1000000}},
+			[]int64{0, 1, 2},
+			[]gaugeChecker{
+				{0, 0, []int64{0}, 0, true, defaultSingleLockDistributedCoins},
+				{1, 1, []int64{1}, 0, true, defaultSingleLockDistributedCoins},
+				{2, 2, []int64{2}, 0, true, defaultSingleLockDistributedCoins}},
+		},
+		// Do not allocate rewards to the Unbonded validator. Therefore gauges are not distributed
+		{
+			"Unallocate to Unbonded validator",
+			[]stakingtypes.BondStatus{stakingtypes.Bonded, stakingtypes.Unbonded},
+			[]superfluidDelegation{{0, 0, 0, 1000000}, {1, 1, 0, 1000000}},
+			[]int64{0},
+			[]gaugeChecker{{0, 0, []int64{0}, 0, true, defaultSingleLockDistributedCoins}, {1, 1, []int64{1}, 0, false, defaultSingleLockDistributedCoins}},
+		},
```

Comment:
A little confused here, why do we allocate rewards to the unbonded validator in the first case, but in the second case we done?

Reply:
@czarcas7ic Are you asking about the reason we don't allocate rewards to the validator here? If so, its so that we don't create gauges. 
We're trying to test both case of having rewarded to the unbonded validator and have not rewarded to the unbonded validator